### PR TITLE
CPU example for NAS RL cifar10 training container

### PR DESCRIPTION
**Pull request by andreyvelich**
I made few improvements in this PR:
1. I added Dockerfile with CPU support for NAS RL training container and fixed problem with parsing current yaml example for NAS job. In the future I will add build of this training container to our CI.
2. Also, I noticed that currently we don't send information about Trials to Suggestion (`ConvertTrials` function return empty Trial set).
Because of that, I added `ConvertTrials ` function to Suggestion client
3. I fixed few bugs in NAS Suggestion. The problem was with running class constructor every `getSuggestions` call.

Please, take a look.
/cc @gaocegege @johnugeorge @hougangliu

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/katib/999)

### CODE REVIEW

File: examples/v1alpha3/nasjob-example-RL.yaml [hunk 1]
```diff
@@ -67,7 +67,7 @@ spec:
                   - "RunTrial.py"
                   {{- with .HyperParameters}}
                   {{- range .}}
-                  - "--{{.Name}}={{.Value}}"
+                  - "--{{.Name}}=\"{{.Value}}\""
```
Comment (johnugeorge) [inline examples/v1alpha3/nasjob-example-RL.yaml]: why is this required?

File: pkg/controller.v1alpha3/suggestion/suggestionclient/suggestionclient.go [hunk 2]
```diff
@@ -174,10 +176,92 @@ func (g *General) ConvertExperiment(e *experimentsv1alpha3.Experiment) *suggesti
 }
 
 // ConvertTrials converts CRD to the GRPC definition.
-func (g *General) ConvertTrials(
-	t []trialsv1alpha3.Trial) []*suggestionapi.Trial {
-	res := make([]*suggestionapi.Trial, 0)
-	return res
+func (g *General) ConvertTrials(ts []trialsv1alpha3.Trial) []*suggestionapi.Trial {
```
Comment (johnugeorge) [inline pkg/controller.v1alpha3/suggestion/suggestionclient/suggestionclient.go]: Strange. Without trial information being sent to suggestion algorithm services, how did it work for other algorithms?
@gaocegege

Comment (gaocegege) [reply pkg/controller.v1alpha3/suggestion/suggestionclient/suggestionclient.go]: We have unit test with trial info. But if the trial info is missing, the algorithm also works. I think it is the reason why we did not find the problem.

Comment (johnugeorge) [reply pkg/controller.v1alpha3/suggestion/suggestionclient/suggestionclient.go]: https://github.com/kubeflow/katib/blob/ebb48f88acb307bcc34f0530a905bcdd456c272f/pkg/suggestion/v1alpha3/skopt_service.py#L30 
 We are using trials from the request. How can bayesianoptimization like algo  work without the previous trial information?

Comment (gaocegege) [reply pkg/controller.v1alpha3/suggestion/suggestionclient/suggestionclient.go]: bayesianoptimization can work since the prior knowledge is always nil. skopt can handle it.

**Comment by johnugeorge** [comment]
grid search has consistently failed. This has to be looked at

**Comment by gaocegege** [comment]
```
2020/01/07 08:16:41 Waiting for Experiment grid-example to finish.
2020/01/07 08:16:41 Experiment grid-example's trials: 2 trials, 0 pending trials,
1 running trials, 0 killed trials, 1 succeeded trials, 0 failed trials.
2020/01/07 08:16:41 Optimal Trial for Experiment grid-example: {grid-example-4wkdd6ww [{--lr 0.002} {--num-layers 2} {--optimizer sgd}] {[{Validation-accuracy 0.966461}]}}
2020/01/07 08:16:41 Experiment grid-example's conditions: [{Created True ExperimentCreated Experiment is created 2020-01-07 08:11:01 +0000 UTC 2020-01-07 08:11:01 +0000 UTC}
{Running True ExperimentRunning Experiment is running 2020-01-07 08:12:20 +0000 UTC 2020-01-07 08:12:20 +0000 UTC}]
2020/01/07 08:16:41 Suggestion grid-example's conditions: [{Created True SuggestionCreated Suggestion is created 2020-01-07 08:11:01 +0000 UTC 2020-01-07 08:11:01 +0000 UTC}
{DeploymentReady True DeploymentReady Deployment is ready 2020-01-07 08:12:12 +0000 UTC 2020-01-07 08:12:12 +0000 UTC} {Running True SuggestionRunning Suggestion is running 2
020-01-07 08:12:12 +0000 UTC 2020-01-07 08:12:12 +0000 UTC}]
2020/01/07 08:16:41 Suggestion grid-example's suggestions: [{[{--lr 0.001} {--num-layers 2} {--optimizer sgd}] grid-example-fm66xcss} {[{--lr 0.002} {--num-layers 2} {--optim
izer sgd}] grid-example-4wkdd6ww}]
2020/01/07 08:17:01 Waiting for Experiment grid-example to finish.
2020/01/07 08:17:01 Experiment grid-example's trials: 2 trials, 0 pending trials,
1 running trials, 0 killed trials, 1 succeeded trials, 0 failed trials.
2020/01/07 08:17:01 Optimal Trial for Experiment grid-example: {grid-example-4wkdd6ww [{--lr 0.002} {--num-layers 2} {--optimizer sgd}] {[{Validation-accuracy 0.966461}]}}
2020/01/07 08:17:01 Experiment grid-example's conditions: [{Created True ExperimentCreated Experiment is created 2020-01-07 08:11:01 +0000 UTC 2020-01-07 08:11:01 +0000 UTC}
{Running True ExperimentRunning Experiment is running 2020-01-07 08:12:20 +0000 UTC 2020-01-07 08:12:20 +0000 UTC}]
2020/01/07 08:17:01 Suggestion grid-example's conditions: [{Created True SuggestionCreated Suggestion is created 2020-01-07 08:11:01 +0000 UTC 2020-01-07 08:11:01 +0000 UTC}
{DeploymentReady True DeploymentReady Deployment is ready 2020-01-07 08:12:12 +0000 UTC 2020-01-07 08:12:12 +0000 UTC} {Running True SuggestionRunning Suggestion is running 2
020-01-07 08:12:12 +0000 UTC 2020-01-07 08:12:12 +0000 UTC}]
2020/01/07 08:17:01 Suggestion grid-example's suggestions: [{[{--lr 0.001} {--num-layers 2} {--optimizer sgd}] grid-example-fm66xcss} {[{--lr 0.002} {--num-layers 2} {--optim
izer sgd}] grid-example-4wkdd6ww}]
```

The hyperparameter is created, but the trials is not finished.

**Comment by johnugeorge** [comment]
We should  try with increased timeout

File: examples/v1alpha3/nasjob-example-RL.yaml (hunk 1 above)
Comment (andreyvelich) [reply examples/v1alpha3/nasjob-example-RL.yaml]: Here https://github.com/andreyvelich/katib/blob/cpu-example-rl-cifar10/examples/v1alpha3/NAS-training-containers/RL-cifar10/RunTrial.py#L13 I parse `architecture` and `nn_config` as a string. The format of architecture something like this `[[15], [86, 1], [7, 1, 0], [57, 1, 0, 0], [3, 0, 1, 1, 0], [42, 1, 1, 1, 1, 0], [17, 0, 0, 0, 1, 0, 0], [49, 0, 1, 1, 0, 0, 0, 1]]` and we need to send it as string. If we pass it without double quotes it fails in parsing.

File: pkg/controller.v1alpha3/suggestion/suggestionclient/suggestionclient.go (hunk 2 above)
Comment (andreyvelich) [reply pkg/controller.v1alpha3/suggestion/suggestionclient/suggestionclient.go]: It is strange why we are trying to convert it from `request.trials`. I faced with the problem when `request.trials` is nil, so it is nothing to convert.

**Comment by andreyvelich** [comment]
I think I found the problem. Here https://chocolate.readthedocs.io/api/sample.html#chocolate.Grid is says that search space to explore can be only discrete dimensions, but we have in our example https://github.com/kubeflow/katib/blob/master/examples/v1alpha3/grid-example.yaml#L33 categorical type also.
Here https://github.com/kubeflow/katib/blob/master/pkg/suggestion/v1alpha3/chocolate/base_chocolate_service.py#L46 we create search space for Grid suggestion with all parameters.
If we remove this categorical parameter from search space for grid, will it be fine?
/cc @gaocegege

**Comment by andreyvelich** [comment]
Hold for now
/hold

**Comment by andreyvelich** [comment]
Tests passed.
/cc @johnugeorge 

/hold cancel

**Comment by gaocegege** [comment]
@andreyvelich Thanks for the fix. It is my fault.

I meet the problem before, but forget that we do not run the unit test for grid search. https://github.com/kubeflow/katib/blob/master/test/suggestion/v1alpha3/test_chocolate_service.py.failed

**Comment by johnugeorge** [comment]
@gaocegege but Is this a acceptable restriction for not allowing to having categorical parameters for grid algorithm?

**Comment by gaocegege** [comment]
@johnugeorge It is not. I think we should investigate if we could solve it.

**Comment by johnugeorge** [comment]
Yeah. I think so.  This restriction looks unacceptable to me as this is one of the common algorithms. 

Should we merge it anyways or wait ?
